### DISCUSSION TITLE
Bug - response not using response_model_by_alias param

### DISCUSSION THREAD

**Discussion by vishalwy** [fastapi/fastapi#8546] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from fastapi.testclient import TestClient
from pydantic import BaseModel, Field


class Data(BaseModel):
    message: str = Field(..., alias="greeting")


app = FastAPI()
data = Data(greeting="Hello World")


@app.get("/", response_model_by_alias=False)
def index():
    return data


client = TestClient(app)


def test_index():
    response = client.get("/")
    assert response.json() == data.dict(by_alias=False), "response should not use alias"


__name__ == "__main__" and test_index()
```

### Description

* Open the browser and call the endpoint `/`. 
* It returns a JSON with `{"greeting":"Hello World"}`.
* But I expected it to return `{"message":"Hello World"}`.

Route `/` is configured with `response_model_by_alias=False`. But in the response, it using alias `greeting` instead of the actual field name `message`.

This line is causing the issue, as it is not forwarding `by_alias` argument. 
https://github.com/tiangolo/fastapi/blob/672c55ac626b20cc6219696023cdcc5871e5141d/fastapi/routing.py#L148 

### Operating System

macOS

### Operating System Details

```
ProductName:	macOS
ProductVersion:	11.6.2
BuildVersion:	20G314
```

### FastAPI Version

0.71.0

### Python Version

3.7.12

**Comment by harunyasar** [comment]
I guess you set it up wrong. I simply recommend you to go through `pydantic` `BaseModel` config [documentation](https://pydantic-docs.helpmanual.io/usage/model_config/).

```python
from fastapi import FastAPI
from fastapi.testclient import TestClient
from pydantic import BaseModel, Field


class Data(BaseModel):
    message: str = Field(..., alias="greeting")

    class Config:
        allow_population_by_field_name = True


app = FastAPI()
data = Data(greeting="Hello World")


@app.get("/", response_model_by_alias=True)
def index():
    return data


client = TestClient(app)


def test_index():
    response = client.get("/")
    assert response.json() == data.dict(by_alias=True), "response should not use alias"

__name__ == "__main__" and test_index()
```

**Comment by harunyasar** [comment]
In fact, I would briefly suggest you define fields in the opposite direction. There are two different discussions on the subject [here](https://github.com/tiangolo/fastapi/issues/1810) and [here](https://github.com/tiangolo/fastapi/issues/771).

**Comment by vishalwy** [comment]
In my code, I am trying to assert `response.json() == data.dict(by_alias=False)`. Here I am comparing the dict representation of `data` returned from the route against `data.dict(by_alias=False)`. Since the route is configured to use `response_model_by_alias=False`, I expect the assert to be successful. But it is failing here since the route is not forwarding `response_model_by_alias` to `jsonable_encoder`.

**Comment by harunyasar** [comment]
@vishalwy I got your case. Have you checked the discussions that I have mentioned above? It seems to be `response_model_by_alias` is discouraged.

**Comment by SiddChugh** [comment]
Can the maintainers verify whether this bug would be addressed or we should keep on using the workaround mentioned [here](https://github.com/tiangolo/fastapi/issues/1810#issuecomment-895126406) ?

**Comment by tares003** [comment]
I have the exact save issue, not consistent behaviour across python versions.

I have this pydantic model and wanting the model to be serialised using the alise,

`class TimeR(BaseModel):
    from_: Optional[str] = Field(..., alias='from')
    to: Optional[str] = None
`


`@router.put("/toggle_pin_it", response_model=schemas.TimeR,  response_model_by_alias=True)
`

**Comment by WuMenghao** [comment]
Code below in module fastapi/encoder.py causes this problem.
```
    if isinstance(obj, BaseModel):
        encoder = getattr(obj.__config__, "json_encoders", {})
        if custom_encoder:
            encoder.update(custom_encoder)
        obj_dict = obj.dict(
            include=include,
            exclude=exclude,
            by_alias=by_alias,
            exclude_unset=exclude_unset,
            exclude_none=exclude_none,
            exclude_defaults=exclude_defaults,
        )
        if "__root__" in obj_dict:
            obj_dict = obj_dict["__root__"]
        return jsonable_encoder(
            obj_dict,
            exclude_none=exclude_none,
            exclude_defaults=exclude_defaults,
            custom_encoder=encoder,
            sqlalchemy_safe=sqlalchemy_safe,
        )
```
if you modify to this will fix this problem
```
    if isinstance(obj, BaseModel):
        encoder = getattr(obj.__config__, "json_encoders", {})
        if custom_encoder:
            encoder.update(custom_encoder)
        obj_dict = obj.dict(
            include=include,
            exclude=exclude,
            by_alias=by_alias,
            exclude_unset=exclude_unset,
            exclude_none=exclude_none,
            exclude_defaults=exclude_defaults,
        )
        if "__root__" in obj_dict:
            obj_dict = obj_dict["__root__"]
        return jsonable_encoder(
            obj_dict,
            include=include,      # add this line
            exclude=exclude,    # add this line
            by_alias=by_alias,   # add this line
            exclude_none=exclude_none,
            exclude_defaults=exclude_defaults,
            custom_encoder=encoder,
            sqlalchemy_safe=sqlalchemy_safe,
        )
```

**Comment by YuriiMotov** [comment, marked as answer]
Actually, the problem here is that you didn't specify `response_model`. In this case FastAPI doesn't validate response and doesn't apply `response_model_by_alias`.

Add response model and everything works (at least in current version):

```python
@app.get("/", response_model_by_alias=False)
def index() -> Data:
    return data
```


IMO, `response_model_by_alias` was bad idea and should be deprecated\removed. It leads to incorrect openapi schema generation.

Use `validation_alias` to solve this:

```python
class Data(BaseModel):
    message: str = Field(..., validation_alias="greeting")
```
